File: dynamic_programming/300_longest_increasing_subsequence.cc
#include <iostream>
#include <vector>

int LengthOfLIS(const std::vector<int>& nums) {
  // dp[i] represents the length of longest increasing subsequence ended with
  // nums[i].
  std::vector<int> dp(nums.size(), 1);
  int length_of_lis = 0;
  for (int i = 0; i < nums.size(); ++i) {
    for (int j = 0; j < i; ++j) {
      if (nums[j] < nums[i]) {
        dp[i] = std::max(dp[i], dp[j] + 1);
      }
    }
    length_of_lis = std::max(length_of_lis, dp[i]);
    std::cout << "dp[" << i << "]: " << dp[i] << std::endl;
  }
  return length_of_lis;
}

int main(int argc, char* argv[]) {
  std::vector<int> nums = {10, 9, 2, 5, 3, 7, 101, 18};
  std::cout << LengthOfLIS(nums) << std::endl;
  return 0;
}

File: dynamic_programming/subarray_and_subsequence/300_longest_increasing_subsequence.py
from typing import List


# 2026/06/08: 自己实现. 这里不设置 dummy 更简单.
def length_of_lis(nums: List[int]) -> int:
    # dp[i] 表示以 i 结尾的最长升序长度.
    dp = [1] * len(nums)
    res = 0
    for i in range(len(nums)):
        for j in range(i):
            if nums[j] < nums[i]:
                dp[i] = max(dp[i], dp[j] + 1)
        res = max(res, dp[i])
    return res


nums = [10, 9, 2, 5, 3, 7, 101, 18]
nums = [1, 3, 6, 7, 9, 4, 10, 5, 6]
print(length_of_lis(nums))
